tianshou/data/batch.py

```diff
@@ -502,49 +546,92 @@ def to_torch(self, dtype: Optional[torch.dtype] = None,
                 v.to_torch(dtype, device)
 
     def cat_(self,
-             batches: Union['Batch', List[Union[dict, 'Batch']]]) -> None:
+             batches: Union['Batch', List[Union[dict, 'Batch']]],
+             lens: Optional[List[int]] = None
+             ) -> None:
         """Concatenate a list of (or one) :class:`~tianshou.data.Batch` objects
         into current batch.
+        If x=Batch(a=Batch(a=np.random.randn(3, 4)), b=np.random.randn(3, 4)),
+        y = Batch(a=Batch(a=Batch()), b=np.random.randn(3, 4)), if we want to
```

Comment:
Reserving keys `a=Batch(a=Batch())` should not be necessary.

Reply:
This is hierarchical key reservation, and is indeed required for tianshou:

> Hierarchical key reservation is a must for the current collector implementation, since it will create something like self.data is a `Batch(obs=..., ..., policy=Batch(_state=Batch()), ...)`

Reply:
That's not right. It should be possible to do this:
```
>>> Batch.stack([Batch(a=np.array([1])), Batch(a=np.array([2]), b=np.array([3]))])
Batch(
    a: array([[1],
              [2]]),
    b: array([[0],
              [3]]),
)
```

This is very likely to append in practice (and it happens in some project of my company). For example, concatenating environment `info`, the keys of the last step (when done flag is True) can be completely different, or a least some new ones can be present, for instance in the case of a terminal reward.

Reply:
> That's not right. It should be possible to do this:

I don't get it. It is already supported now.

Reply:
So the doc is wrong. Discussed in the related issue.

Reply:
Solved in #139 .

Reply:
Mark resolved after fixing doc.

Reply:
The doc is this [comment](https://github.com/thu-ml/tianshou/issues/139#issuecomment-659092494) is already fixed.

Reply:
Sure, but is has to be fixed in this PR, line 137.

Reply:
Oh, my bad. Fixed in b795493 .